source/smart-contracts/guides/custom-errors.rst

```diff
@@ -0,0 +1,55 @@
+.. _custom-errors:
+
+====================
+Return custom errors
+====================
+
+This guide shows how to return custom errors from your Rust smart contract.
+
+Defining and deriving
+=====================
+
+Custom error codes help communicate why a contract rejects and can be returned
+both during initialization and during updates.
+
+On-chain, smart contracts return a numeric error code when rejecting. This is
+also the case, when using a custom error type. Therefore, a mapping from the
+custom error type to ``Reject``, in the form of an implementation of
+``From<MyError> for Reject``, is needed. We can also derive it
+automatically using ``#[derive(Reject)]``::
+
+   #[derive(Reject)]
+   enum MyError {
+       ErrOne,
+       ErrTwo,
+   }
+
+.. note::
+
+   The valid range of error codes is ``i32::MIN..-1``. When deriving
+   ``Reject``, each variant is assigned an error code as determined by the
+   ordering. First variant (``ErrOne`` in example) gets ``-1``, second variant
+   (``ErrTwo`` in example) gets ``-2``, and so on.
+
+.. warning::
+
+   Deriving ``Reject`` is only possible for fieldless enums, i.e., enums where
+   the variants do not have associated data.
```

Comment:
Maybe also point out they should not have custom discriminants, e.g., they should not have things like
```rust
enum Foo {
   Foo = 3,
   Bar = 7,
}